docs: include a section with CircleCI instructions (#318)
game-ci/documentation: docs/11-circleci/03-build.mdx

---
toc_max_heading_level: 5

> ---

Author: Done ✔️

# Build

The Unity orb provides the
[build job](https://circleci.com/developer/orbs/orb/game-ci/unity#jobs-build) to facilitate the
process of building your Unity project to multiple platforms.

### Parameters

> In this sub-section, you will find information on each parameter the job provides.

Author: Done ✔️

> In this sub-section, you will find information on each parameter the job provides.

Reviewer: By the way, you can commit suggestions right from github by using the `commit suggestions` or `add suggestion to batch` buttons. That way it automatically references the change to the thread.

> In this sub-section, you will find information on each parameter the job provides.

Author: Oh yeah, never used that before. Good to know! 100% using it in the next iteration.

> #### build-name

Reviewer: I'm slightly confused by the usage of kebab-case for parameters.

In previous implementations we used camelCase, whereas CircleCI usually uses snake_case (or maybe I'm mistaken?).

It's fine but just curious if this is based on something.

Author: > CircleCI usually uses snake_case (or maybe I'm mistaken?).

That's correct. Parameters in CircleCI use snake_case. However, we've been using kebab for parameters in official orbs for the longest time. I don't have that decision's context, though.

> - _Default_: `UnityBuild`

Author: Do you mean to use the repo name as the value for this parameter if it's left empty?

> - _Default_: `UnityBuild`

Reviewer: I thought we were passing no build name, and Unity would figure it out by itself.

However this is what our logic really looks like in `unity-builder` ([ref](https://github.com/game-ci/unity-builder/blob/c15f955669bd556aae56eb5e5425a263ac058c9c/src/model/input.ts#L112)):

![image](https://user-images.githubusercontent.com/20756439/186034612-83cb5326-36cd-4674-b7b5-75a4f29b83ad.png)

> - _Default_: `UnityBuild`

Author: Yea, the build name is a required parameter. The execute method I'm using in the orb fails if the parameter comes empty ([ref](https://github.com/game-ci/unity-orb/blob/main/src/dependencies/unity-builder/BuildCommand.cs#L76)):

```c#
static string GetBuildName()
{
    string buildName = GetArgument("customBuildName");
    Console.WriteLine(":: Received customBuildName " + buildName);
    if (buildName == "")
    {
        throw new Exception("customBuildName argument is missing");
    }
    return buildName;
}
```

> - _Default_: `UnityBuild`

Reviewer: Yea, that's why we're falling back to targetPlatform (which is the only required parameter besides credentials in GitHub).

It would be nice if not everyone had to put a build name if they didn't want to. Target platform could be a good default here. (and if you wanna use the CLI later, you'd not have a breaking change if you make it like that now already)

> - _Default_: `UnityBuild`

Author: Done ✔️

- _Required_: `False`

> - _Type_: `String`

Author: Done ✔️

Your build's name. If you choose to compress your artifacts, this name will be used in the
compressed file.

```yaml
version: 2.1

orbs:
  unity: game-ci/unity@x.y

workflows:
  build-unity-project:
    jobs:
      - unity/build:
          build-name: 'MyUnityProject'
```

#### build-target

- _Default_: `N/A`
- _Required_: `True`
- _Type_: `String`

> The platform on which you want to run your game. Valid parameters can be found in the Unity Editor
> [documentation](https://docs.unity3d.com/2022.2/Documentation/ScriptReference/BuildTarget.html).

Reviewer: ```suggestion
The platform on which you want to run your game or application. Available targets can be found in the Unity Editor
[documentation](https://docs.unity3d.com/2022.2/Documentation/ScriptReference/BuildTarget.html).
```

Reviewer: Quick note: even though we're called GameCI, we acknowledge that Unity is used for many use cases outside of games, like educational projects, 3d guided tours and many more.

Author: That's fair. And done ✔️

```yaml
version: 2.1

orbs:
  unity: game-ci/unity@x.y

workflows:
  build-unity-project:
    jobs:
      - unity/build:
          build-target: 'StandaloneLinux64'
```

#### compress

- _Default_: `true`
- _Required_: `False`
- _Type_: `Boolean`

> Whether to compress the build output to a `.tar.gz` file. Set it to `true` if you want to download
> the build artifacts from the CircleCI dashboard, otherwise, you will have to download each file
> individually. If set to `false` for
> [decompressed WebGL](https://docs.unity3d.com/Manual/webgl-deploying.html) builds, the built project
> can be run directly from the GameCI dashboard.

Reviewer: ```suggestion
Whether to compress the build output to a `.tar.gz` file. Set it to `true` if you want to download
the build artifacts from the CircleCI dashboard. Otherwise, you will have to download each file
individually. If set to `false` for
[decompressed WebGL](https://docs.unity3d.com/Manual/webgl-deploying.html) builds, the built project
can be run directly from the GameCI dashboard.
```

Author: Done ✔️

```yaml
version: 2.1

orbs:
  unity: game-ci/unity@x.y

workflows:
  build-unity-project:
    jobs:
      - unity/build:
          compress: true
```

#### executor

- _Default_: `N/A`
- _Required_: `True`
- _Type_: `Executor`

> The executor on which you want to run your build job on. Bear in mind that projects using `IL2CPP`
> as the [scripting backend](https://docs.unity3d.com/Manual/scripting-backends.html) should run on
> the same OS as the [build target](#build-target). For example, if you want to build for macOS using
> `IL2CPP`, you must use the
> [macos](https://circleci.com/developer/orbs/orb/game-ci/unity#executors-macos) or
> [macos-runner](https://circleci.com/developer/orbs/orb/game-ci/unity#executors-macos-runner)
> executors.

Reviewer: ```suggestion
The executor that you want to run your build job on. Bear in mind that builds using `IL2CPP` for a 
[scripting backend](https://docs.unity3d.com/Manual/scripting-backends.html) should run on
the same OS as the [build target](#build-target). 

For example, if you want to build for macOS using `IL2CPP`, then you must use the
[macos](https://circleci.com/developer/orbs/orb/game-ci/unity#executors-macos) or
[macos-runner](https://circleci.com/developer/orbs/orb/game-ci/unity#executors-macos-runner)
executor.
```

Author: Done ✔️

Author: After this change, I reread this section and believe we should use the original. The second sentence is a direct continuation of the first, but making it a paragraph implies otherwise and breaks the reading flow.

Notice that the Windows and macOS executors have a longer environment setup time than
[ubuntu](https://circleci.com/developer/orbs/orb/game-ci/unity#executors-ubuntu). This is due to the
installation of the Windows SDK in `Windows` and the installation of the Unity Hub and Unity Editor
in `macOS`. Should you require faster build times in these executors, consider using their

> non-ephemeral runner counterparts.

Reviewer: Let's leave this explanation on the executors page. Concise docs are easier to maintain and will be read more carefully.

Author: Done ✔️

Each executor has its set of parameters. See the [Executors](/docs/circleci/executors) section for
more details.

```yaml
version: 2.1

orbs:
  unity: game-ci/unity@x.y

workflows:
  build-unity-project:
    jobs:
      - unity/build:
          executor:
            name: 'unity/ubuntu'
            target_platform: 'linux-il2cpp'
            editor_version: '2021.3.2f1'
            resource_class: 'large'
```

#### project-path

- _Default_: `.`
- _Required_: `False`
- _Type_: `String`

> If your project sits at the root of your repository, leave it to the default value. Otherwise, enter
> the path of your Unity project. This should be the directory that has an "Assets" folder inside it.

Reviewer: ```suggestion
If your project sits at the root of your repository, leave it to the default value. Otherwise, enter
the path to your Unity project. It should be the directory that has the "Assets" folder inside it.
```

```yaml
version: 2.1

orbs:
  unity: game-ci/unity@x.y

workflows:
  build-unity-project:
    jobs:
      - unity/build:
          project-path: 'Unity2D-Demo-Game-CI-CD/src'
```

#### step-name

- _Default_: `Build the project`
- _Required_: `False`
- _Type_: `String`

> Specify a custom step name to be shown in the CircleCI dashboard.

Reviewer 2: ```suggestion
Specify a custom step name to be shown in the CircleCI web app.
```

```yaml
version: 2.1

orbs:
  unity: game-ci/unity@x.y

workflows:
  build-unity-project:
    jobs:
      - unity/build:
          step-name: 'Build my Unity project'
```

#### unity-license-var-name

- _Default_: `UNITY_ENCODED_LICENSE`
- _Required_: `False`
- _Type_: `env_var_name`

> The name of the environment variable holding the license set in your
> [Context](/docs/circleci/activation#preparing-the-context). If you are building your project using
> the Unity Pro or Plus license, this is **not** required. See [Activation](/docs/circleci/activation)
> for more details.

Reviewer: ```suggestion
The name of the environment variable holding the license set in your
[Context](/docs/circleci/activation#preparing-the-context). Only required when using a Personal License. See [activation](/docs/circleci/activation)
for more details.
```

Reviewer: Lower case because activation is not a keyword.

```yaml
version: 2.1

orbs:
  unity: game-ci/unity@x.y

workflows:
  build-unity-project:
    jobs:
      - unity/build:
          unity-license-var-name: 'UNITY_ENCODED_LICENSE'
```

#### unity-password-var-name

- _Default_: `UNITY_PASSWORD`
- _Required_: `False`
- _Type_: `env_var_name`

> The name of the environment variable holding the password set in your
> [Context](/docs/circleci/activation#preparing-the-context). See
> [Activation](/docs/circleci/activation) for more details.

Reviewer: ```suggestion
The name of the environment variable holding the password set in your
[Context](/docs/circleci/activation#preparing-the-context). See
[activation](/docs/circleci/activation) for more details.
```

```yaml
version: 2.1

orbs:
  unity: game-ci/unity@x.y

workflows:
  build-unity-project:
    jobs:
      - unity/build:
          unity-password-var-name: 'UNITY_PASSWORD'
```

#### unity-serial-var-name

- _Default_: `UNITY_SERIAL`
- _Required_: `False`
- _Type_: `env_var_name`

> The name of the environment variable holding the serial set in your
> [Context](/docs/circleci/activation#preparing-the-context). If you are building your project using
> the Unity Personal license, this is **not** required. See [Activation](/docs/circleci/activation)
> for more details.

Reviewer: ```suggestion
The name of the environment variable holding the serial set in your
[Context](/docs/circleci/activation#preparing-the-context). Only required when using a Pro or Plus License. 
See [activation](/docs/circleci/activation) for more details.
```

```yaml
version: 2.1

orbs:
  unity: game-ci/unity@x.y

workflows:
  build-unity-project:
    jobs:
      - unity/build:
          unity-serial-var-name: 'UNITY_SERIAL'
```

#### unity-username-var-name

- _Default_: `UNITY_USERNAME`
- _Required_: `False`
- _Type_: `env_var_name`

The name of the environment variable holding the username set in your
[Context](/docs/circleci/activation#preparing-the-context). See
[Activation](/docs/circleci/activation) for more details.

```yaml
version: 2.1

orbs:
  unity: game-ci/unity@x.y

workflows:
  build-unity-project:
    jobs:
      - unity/build:
          unity-username-var-name: 'UNITY_USERNAME'
```

### Examples

> In this sub-section, you will find examples of workflows using this job. However, notice that the
> separation used in the examples is for visualization's sake only. You can build for all target
> platforms in the same workflow.

Reviewer: ```suggestion
In this section, you will find examples of workflows using this job. However, notice that the
separation used in the examples is only for the sake of visualisation. You can build for all target
platforms in the same workflow.
```

> 

Reviewer: ```suggestion
All examples are based on CircleCI's
[Demo Project](https://github.com/CircleCI-Public/Unity2D-Demo-Game-CI-CD).

```

#### Personal License

> Below you will see how the build job can be used to build your project for several platforms using
> Unity's personal license. If you are unsure what this means, navigate to the
> [Activation](/docs/circleci/activation) section. All examples are based on CircleCI's
> [Demo Project](https://github.com/CircleCI-Public/Unity2D-Demo-Game-CI-CD).

Reviewer: ```suggestion
This examples shows how to build your project for several platforms using a Personal License. 
```

Author: I'm just replacing `examples` with `example`. And done ✔️

          build-target: 'Android'
          context: unity

      # iOS
      - unity/build:
          name: 'build-ios'
          step-name: 'Build iOS'
          unity-license-var-name: 'UNITY_ENCODED_LICENSE'
          unity-username-var-name: 'UNITY_USERNAME'
          unity-password-var-name: 'UNITY_PASSWORD'
          executor:
            name: 'unity/ubuntu'
            target_platform: 'ios'
            editor_version: '2021.3.2f1'
            resource_class: 'large'
          project-path: 'Unity2D-Demo-Game-CI-CD/src'
          build-target: 'iOS'
          context: unity

      # tvOS
      - unity/build:
          name: 'build-tvOS'
          step-name: "Build Apple's tvOS"
          unity-license-var-name: 'UNITY_ENCODED_LICENSE'
          unity-username-var-name: 'UNITY_USERNAME'
          unity-password-var-name: 'UNITY_PASSWORD'
          executor:
            name: 'unity/windows-2019'
            size: 'large'
            editor_version: '2021.3.2f1'
            target_platform: 'appletv'
          project-path: 'Unity2D-Demo-Game-CI-CD/src'
          build-target: tvOS
          compress: true
          context: unity
```

#### Professional License

> Below you will see how the build job can be used to build your project for several platforms using
> Unity's professional license. If you are unsure what this means, navigate to the
> [Activation](/docs/circleci/activation) section. All examples are based on CircleCI's
> [Demo Project](https://github.com/CircleCI-Public/Unity2D-Demo-Game-CI-CD).

Reviewer: ```suggestion
This examples shows how to build your project for several platforms using a Pro License.
```

Reviewer: Be careful not to copy paste too much, repetitive texts cause people to stop reading what you write and it's very prone to bugs.

Author: Same as above.
